gateway/src/traits.rs

```diff
@@ -4,26 +4,159 @@
-use frame_support::weights::{constants::RocksDbWeight, Weight};
+use crate::Config;
+use crate::Error::ErrorForwarding;
+use codec::Decode;
+use frame_support::dispatch::{DispatchResult, RawOrigin};
+use frame_support::weights::Weight;
+use sp_core::H160;
+use sp_runtime::{traits::Dispatchable, DispatchError};
+use xcm::latest::prelude::*;
+use xcm::latest::Xcm;
+use xcm::prelude::DescendOrigin;
 
 // ----------------------------------------------------------------------------
 // Traits declaration
 // ----------------------------------------------------------------------------
 
+pub const GATEWAY_PALLET_ID: PalletId = PalletId(*b"axgteway");
+
 /// Weight information for pallet extrinsics
 ///
 /// Weights are calculated using runtime benchmarking features.
 /// See [`benchmarking`] module for more information.
 pub trait WeightInfo {
-    fn transfer_operatorship() -> Weight;
+    fn execute(c: u32) -> Weight;
+    fn transfer_operatorship(c: u32) -> Weight;
+    fn approve_contract_call() -> Weight;
+    fn forward_approved_call() -> Weight;
 }
 
 // For backwards compatibility and tests
 impl WeightInfo for () {
-    fn transfer_operatorship() -> Weight {
-        (Weight::from_ref_time(65_453_000))
-            .saturating_add(RocksDbWeight::get().reads(3_u64))
-            .saturating_add(RocksDbWeight::get().writes(1_u64))
+    fn execute(c: u32) -> Weight {
+        // Minimum execution time: 13_885 nanoseconds.
+        Weight::from_ref_time(20_147_978 as u64)
+            // Standard Error: 2_232
+            .saturating_add(Weight::from_ref_time(3_516_969 as u64).saturating_mul(c as u64))
+    }
+    fn transfer_operatorship(c: u32) -> Weight {
+        // Minimum execution time: 14_470 nanoseconds.
+        Weight::from_ref_time(17_443_346 as u64)
+            // Standard Error: 2_037
+            .saturating_add(Weight::from_ref_time(3_510_555 as u64).saturating_mul(c as u64))
+    }
+    fn approve_contract_call() -> Weight {
+        Weight::from_ref_time(17_443_346 as u64)
+    }
+    fn forward_approved_call() -> Weight {
+        Weight::from_ref_time(17_443_346 as u64)
+    }
+}
+
+pub trait ApprovedCallForwarder<AccountId, T> {
+    fn is_local() -> bool;
+    fn do_forward(
+        who: AccountId,
+        source_chain: String,
+        source_address: String,
+        contract_address: H160,
+        dest: MultiLocation,
+        call: Vec<u8>,
+    ) -> DispatchResult;
+}
+
+/// Local Forwarder Default Implementation
+pub struct ApprovedCallLocalForwarder;
+impl<T: Config> ApprovedCallForwarder<T::AccountId, T> for ApprovedCallLocalForwarder {
+    fn is_local() -> bool {
+        true
+    }
+
+    fn do_forward(
+        who: T::AccountId,
+        _source_chain: String,
+        _source_address: String,
+        _contract_address: H160,
+        _dest: MultiLocation,
+        call: Vec<u8>,
+    ) -> DispatchResult {
+        match <T as Config>::RuntimeCall::decode(&mut &call[..]) {
```

Comment:
This can also be written like this:
```suggestion
        match RuntimeCall::<T>::decode(&mut &call[..]) {
```

Reply:
It doesnt seem to compile